Optimize twisted.trial.unittest.TestCase

p12tic opened the pull request:
This PR introduces a several optimizations for twisted.trial.unittest.TestCase. The optimizations are best reviewed one by one.

As part of optimizations, reactor stop, iterate and crash now raise exception instead of a warning. Calling these functions from tests has been deprecated 18 years ago in 40dcfe64dc3cb3dfd768b1ef2cbf56093c9a9631.

p12tic committed:
Only call runWithWarningsSuppressed() if there are suppressions

p12tic committed:
Use separate members for reactor methods that are deprecated in tests

p12tic committed:
Simplify access to deprecated reactor methods in TestCase

p12tic committed:
Use global variable directly in TestCase._wait()

p12tic committed:
Simplify reactor method override handling in TestCase

p12tic committed:
Simplify _Janitor._cleanPending to access reactor.iterate() directly

This also speeds up test execution slightly because there are less usage
of utils.suppressWarnings().

p12tic committed:
Add newsfragment

exarkun commented on src/twisted/trial/_asynctest.py:
```diff
@@ -139,8 +112,39 @@ def onTimeout(d):
 
-        call = self._twistedPrivateScheduler.callLater(timeout, onTimeout, d)
-        d.addBoth(lambda x: call.active() and call.cancel() or x)
+        if d.called or d.callbacks:
```
No - this isn't how you use Deferred.

p12tic replied:
That's why I marked the PR as draft :)

However, could we expose `Deferred.hasResult()` or something like that which would give the caller information about whether there are any pending callbacks or not?

exarkun replied:
Unfortunately, no.  It's fundamentally incompatible with the way Deferred works.  The way you get a result is you add a callback.  Any other approach risks receiving a different outcome than the APIs are supposed to produce.  In an application you might decide to accept this risk (perhaps because you know that in some case in your application it cannot result in a problem) but in trial where arbitrary application code will produce the Deferreds, it's not reasonable.

glyph replied:
It is absolutely a mistake that we exposed `.called` and `.callbacks` as public attributes in the first place, and something we should eventually correct.

p12tic replied:
I fully agree that accessing `called` and `callbacks` is bad idea. The idea with `hasResult()` is to allow the caller to understand whether a adding an additional callback would result in it being called synchronously. User would still need to actually add a callback to fetch the result.

I expect that roughly ~half~ fifth of the 30% performance improvement is due to having an equivalent of `hasResult()` and not doing additional unnecessary work.

p12tic replied:
I've removed the change and the performance optimization has been reduced from 30% to 19% as tested by codspeed.

p12tic committed:
Track deprecated reactor methods at global scope

Some tests initiate multiple TestCase instances that all run at the same
time, so reactor method deprecation must be tracked at global scope.

p12tic committed:
Fail test when deprecated crash, iterate or stop are called on reactor

p12tic committed:
Make _waitIsRunning bool instead of a list

p12tic committed:
Refactor results variable from list to bool in TestCase._run()

p12tic committed:
Add test for edge case when test expects failure during timeout

Currently there is a test for the case when the returned Deferred does
not have `callback()` called. However, there is no test for the case
when `callback()` is called, but the callback chain does not finish
before timeout.
